### DISCUSSION TITLE
How to best reference the current value of an atom?

### DISCUSSION THREAD

**Discussion by wdoug** [pmndrs/jotai#1758] [General] [answered]
I am trying to load some data with `useEffect` in a reusable hook but I only want to load that data once on initial load. I tried storing the state of whether that data has already been loaded in an atom, but I don't see a nice way to properly reference its value where I need to.

Here is a much-simplified version of my initial naive code:
```ts
const isRequestedAtom = atom(false);

const useData = () => {
  const [isRequested, setIsRequested] = useAtom(isRequestedAtom);

  useEffect(() => {
    if (!isRequested) {
      setIsRequested(true);
      console.log("This should only be logged once"); // but it isn't
      fetchData();
    }
  }, [isRequested, setIsRequested]);
};
```
That doesn't work when `useData` is called multiple times because the `isRequested` value in the `useEffect` is the stale value from the above `useAtom` call. 

I was able to get it working with this, but it feels like an inappropriate hack:
```ts
const isRequestedRefAtom = atom({ current: false });

const useData2 = () => {
  const isRequestedRef = useAtomValue(isRequestedRefAtom);

  useEffect(() => {
    if (!isRequestedRef.current) {
      isRequestedRef.current = true;
      console.log("This should only be logged once"); // this works
      fetchData();
    }
  }, [isRequestedRef]);
};
```

After searching through different issues and docs, I found about the [useAtomCallback](https://jotai.org/docs/utilities/callback#useatomcallback) which seemed like it might work for this use case, but I couldn't figure it out if it does.

I also saw [this suggestion](https://github.com/pmndrs/jotai/discussions/1586) for a `useAtomRef` hook. As suggested in that discussion, I could probably use the new `createStore` API, but that seems like overkill for what I would think would be a basic use case like this one.

What would be a recommended approach for something like this?

**Comment by dai-shi** [comment, marked as answer]
useAtomCallback is one solution.

Personally, I'd prefer write-only atom like this.

```js
const isRequestedAtom = atom(false);

const fetchDataAtom = atom(null, (get, set) => {
  const isRequested = get(isRequestedAtom);
  if (!isRequested) {
    set(isRequestedAtom, true);
    console.log("This should only be logged once");
    fetchData();
  }
});

const useData = () => {
  const doFetchData = useSetAtom(fetchDataAtom);
  useEffect(() => {
    doFetchData();
  }, [doFetchData]);
};
```

**Reply by wdoug** [reply]
What would the recommendation be if I wanted to also pass other arguments through to determine which data to request and how to do it? For example, if I am requesting and caching data by id. 

If I went back to the hacky `useData2` example, I could do something like this:
```js
const isRequestedRefAtom = atom({ current: {} });

const useData2 = (id, otherArguments) => {
  const isRequestedRef = useAtomValue(isRequestedRefAtom);

  useEffect(() => {
    if (!isRequestedRef.current[id]) {
      isRequestedRef.current[id] = true;
      fetchData(id, otherArguments);
    }
  }, [isRequestedRef]);
};
```

**Reply by dai-shi** [reply]
You can use `useAtomCallback` in such case.

I'd still prefer write-only atom, which should be defined in useMemo in this case. And, I'd avoid "ref".
```js
const isRequestedAtom = atom({});

const useData2 = (id, otherArguments) => {
  const fetchDataAtom = useMemo(() => atom(null, (get, set) => {
    const isRequested = get(isRequestedAtom);
    if (!isRequested[id]) {
      set(isRequestedAtom, (prev) => ({ ...prev, [id]: true }));
      fetchData(id, otherArguments);
    }
  }), [id, otherArguments]);
  const doFetchData = useSetAtom(fetchDataAtom);
  useEffect(() => {
    doFetchData();
  }, [doFetchData]); 
};
```
